src/plonk/lookup/verifier.rs

```diff
@@ -0,0 +1,79 @@
+use super::super::ConstraintSystem;
+use super::{Lookup, Proof};
+use crate::arithmetic::{CurveAffine, Field};
+
+impl<C: CurveAffine> Proof<C> {
+    /// Given the evaluations of the input columns and table columns of a
+    /// Lookup, along with the evaluations of the permuted input column,
+    /// permuted table column, and lookup grand product, this method applies
+    /// certain constraints to these evaluations and returns the final
+    /// values as a result of applying each constraint.
+    pub fn evaluate_lookup_constraints(
+        &self,
+        cs: &ConstraintSystem<C::Scalar>,
+        beta: C::Scalar,
+        gamma: C::Scalar,
+        theta: C::Scalar,
+        l_0: C::Scalar,
+        lookup: &Lookup,
+        advice_evals: &[C::Scalar],
+        fixed_evals: &[C::Scalar],
+    ) -> Vec<C::Scalar> {
+        let mut constraints = Vec::with_capacity(4);
+        // l_0(X) * (1 - z'(X)) = 0
+        {
+            let first_product_constraint = l_0 * &(C::Scalar::one() - &self.product_eval);
+            constraints.push(first_product_constraint);
+        }
+
+        // z'(X) (a'(X) + \beta) (s'(X) + \gamma)
+        // - z'(\omega^{-1} X) (a_1(X) + \theta a_2(X) + ... + \beta) (s_1(X) + \theta s_2(X) + ... + \gamma)
+        {
+            let left = self.product_eval
+                * &(self.permuted_input_eval + &beta)
+                * &(self.permuted_table_eval + &gamma);
+
+            let mut right = self.product_inv_eval;
+            let mut input_term = C::Scalar::zero();
+            for &input in lookup.input_columns.iter() {
+                let eval = advice_evals[cs.get_any_query_index(input, 0)].clone();
```

Comment:
This code is implicitly assuming that all of the "input" columns are advice columns because it's taking the index from `get_any_query_index` and using it regardless of `input`'s column type.